### DISCUSSION TITLE
Stores with React Context: Memory Leak Issue or Redux DevTools Bug?

### DISCUSSION THREAD

**Discussion by frin88** [pmndrs/zustand#2540] [Q&A] [answered]
Hello there,
I'm using Zustand in a React project written in typescript.

I use `zustand/middleware` and  [Redux DevTools](https://chromewebstore.google.com/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?pli=1) chrome extension to inspect the stores.

I have a parent component that renders a list of items. For each item, I create on the a store provider using react contex by following [this pattern](https://docs.pmnd.rs/zustand/guides/initialize-state-with-props#creating-a-context-with-react.createcontext) from docs

```<Parent>
  {items.map(item => (
    <StoreProvider item={item}>
      <ItemComponent />
    </StoreProvider>
  ))}
<Parent>
```

When the <Parent> component is unmounted the ReduxDevTools inspector still show all previously created stores.
 I've also implemented a quick logic in the parent to keep track of the created stores and destroy them on unmount:

something like

```
// Cleanup function in ParentComponent
useEffect(() => {
  return () => {
    storesTracker.forEach(store => {
      store.destroy();
    });
  };
}, []);
```

So, my question is twofold: Are these ghost stores a sign of memory leaks or just a glitch/bug in Redux DevTools? And do I really need to worry about destroying stores, or is it more of a nice-to-have cleanup?
Am I missing something?

Any insights or advice would be greatly appreciated!

**Comment by dbritto-dev** [comment]
@frin88 as I know it's a known issue with devtools middleware but may I'm wrong

@dai-shi do you have any thoughts?

**Reply by dai-shi** [reply]
@frin88 If you don't use Redux DevTools and still find memory leaks, that sounds bad. Otherwise, it should be fine.

**Reply by frin88** [reply]
@dai-shi  thanks for the response! Just two quick follow-up questions:

1. So, if I've got this right, I can safely skip the step of destroying stores, correct?
2. Also, aside from Redux DevTools, are there any other tools you'd recommend for inspecting stores?

**Reply by dai-shi** [reply, marked as answer]
Yes, stores are garbage collectable. I would recommend to make sure all listeners are unsubscribed.
I'm not sure if we have stable devtools other than redux one.
https://github.com/pmndrs/zustand/blob/9d24d11e1e37c74f64e5c4ccd1c461d6c23522a7/docs/integrations/third-party-libraries.md?plain=1#L37
seems like an experimental one.
I think it's open space for anyone to jump in and develop Zustand-specific devtools.
FYI, Jotai has one: https://github.com/jotaijs/jotai-devtools
